tools/mpy_ld.py

```diff
@@ -590,40 +670,75 @@ def do_relocation_text(env, text_addr, r):
+        # with no offsets to fix up.
+        return
+
+    elif env.arch.name == "EM_RISCV":
+        (addr, value) = process_riscv32_relocation(env, text_addr, r)
+
     else:
         # Unknown/unsupported relocation
         assert 0, r_info_type
 
-    # Write relocation
-    if reloc_type == "le32":
-        (existing,) = struct.unpack_from("<I", env.full_text, r_offset)
-        struct.pack_into("<I", env.full_text, r_offset, (existing + reloc) & 0xFFFFFFFF)
-    elif reloc_type == "thumb_b":
-        b_h, b_l = struct.unpack_from("<HH", env.full_text, r_offset)
-        existing = (b_h & 0x7FF) << 12 | (b_l & 0x7FF) << 1
-        if existing >= 0x400000:  # 2's complement
-            existing -= 0x800000
-        new = existing + reloc
-        b_h = (b_h & 0xF800) | (new >> 12) & 0x7FF
-        b_l = (b_l & 0xF800) | (new >> 1) & 0x7FF
-        struct.pack_into("<HH", env.full_text, r_offset, b_h, b_l)
-    elif reloc_type == "xtensa_l32r":
-        l32r = unpack_u24le(env.full_text, r_offset)
-        assert l32r & 0xF == 1  # RI16 encoded l32r
-        l32r_imm16 = l32r >> 8
-        l32r_imm16 = (l32r_imm16 + reloc >> 2) & 0xFFFF
-        l32r = l32r & 0xFF | l32r_imm16 << 8
-        pack_u24le(env.full_text, r_offset, l32r)
-    else:
-        assert 0, reloc_type
+    if env.arch.name != "EM_RISCV":
+        # Write relocation
+        if reloc_type == "le32":
+            (existing,) = struct.unpack_from("<I", env.full_text, r_offset)
+            struct.pack_into("<I", env.full_text, r_offset, (existing + reloc) & 0xFFFFFFFF)
+        elif reloc_type == "thumb_b":
+            b_h, b_l = struct.unpack_from("<HH", env.full_text, r_offset)
+            existing = (b_h & 0x7FF) << 12 | (b_l & 0x7FF) << 1
+            if existing >= 0x400000:  # 2's complement
+                existing -= 0x800000
+            new = existing + reloc
+            b_h = (b_h & 0xF800) | (new >> 12) & 0x7FF
+            b_l = (b_l & 0xF800) | (new >> 1) & 0x7FF
+            struct.pack_into("<HH", env.full_text, r_offset, b_h, b_l)
+        elif reloc_type == "xtensa_l32r":
+            l32r = unpack_u24le(env.full_text, r_offset)
+            assert l32r & 0xF == 1  # RI16 encoded l32r
+            l32r_imm16 = l32r >> 8
+            l32r_imm16 = (l32r_imm16 + reloc >> 2) & 0xFFFF
+            l32r = l32r & 0xFF | l32r_imm16 << 8
+            pack_u24le(env.full_text, r_offset, l32r)
+        else:
+            assert 0, reloc_type
 
     # Log information about relocation
     if log_name is None:
         if s_type == "STT_SECTION":
             log_name = s.section.name
         else:
             log_name = s.name
-    log(LOG_LEVEL_3, "  {:08x} {} -> {:08x}".format(r_offset, log_name, addr))
+    if addr is not None:
+        log(LOG_LEVEL_3, "  {:08x} {} -> {:08x}".format(r_offset, log_name, addr))
+    else:
+        log(LOG_LEVEL_3, "  {:08x} {} {:08x}".format(r_offset, log_name, value))
```

Comment:
Maybe `"  {:08x} {} == {:08x}"`?

Reply:
This should be handled in the latest revision of this PR.  Thanks for the suggestion!